### PR TITLE
Feature/reflection based dependency injection (#3)

### PR DESCRIPTION
**Pull request by felixsand**
* Updated dependencies to PHP >=7.1 and Symfony ^4.0
Updated starlit inter-package dependencies

* Updated Travis CI to run tests on PHP 7.1 and 7.2

* Fixed typo

* First version of reflection-based DI Container

* Readability and style fixes

* Refactored code to use new version of DI Container
Modified tests affected by code changes

* Code style fix

* Fixed incorrect parameter resolution

* Fixed resolution of classes without constructors

* Fixed resolution of classes without constructors

* Fixed resolution of unmapped interface parameters with default values

* Slight readability improvements

* Slight optimizations and improvements

* Slight optimizations and restructuring of code

* Removed fully qualified function calls

* Simplified test with mock object and expectations

* Added interfaces for Router and View
Changed class name strings into SomeClass::class constructs

* Added unset() and unalias() methods to Container

* Changed uses of Router to RouterInterface

* Allow null parameter to View::escape()

* Added alias check to Container::destroyInstance()

* Fixed test

### CODE REVIEW

**Comment by felixsand** [comment]
Some remaining polishing remains. See: https://github.com/starweb/starlit-app/pull/3

File: src/ViewInterface.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
+<?php
+/**
+ * Starlit App.
+ *
+ * @copyright Copyright (c) 2016 Starweb AB
```
Comment (felixsand) [inline src/ViewInterface.php]:
```suggestion
 * @copyright Copyright (c) 2019 Starweb AB
```
Both here and in any new file

File: src/ViewInterface.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Starlit\App\ViewHelper\AbstractViewHelper;
+
+/**
+ * Abstract implementation of a View
+ */
+interface ViewInterface
+{
+    /**
+     * Constructor.
```
Comment (felixsand) [inline src/ViewInterface.php]: Many unnecessary comments in this file. Would be nice to keep it a bit more slim and only comment where something is unclear or requires some additional info

File: src/Container/Container.php [hunk 3]
```diff
@@ -36,76 +41,131 @@ class Container implements ContainerInterface
...
+
+        return isset($this->dicValues[$key]);
     }
 
     /**
      * @param string $key
      * @return bool
      */
-    public function hasInstance($key)
+    public function hasInstance($key): bool
```
Comment (lordrhodos) [inline src/Container/Container.php:95]:
```suggestion
    public function hasInstance(string $key): bool
```

and phpdoc can be removed

File: src/Container/Container.php [hunk 4]
```diff
@@ -36,76 +41,131 @@ class Container implements ContainerInterface
...
-
-            return $this->dicObjects[$key];
+        try {
+            if (isset($this->dicValues[$key])) {
+                $instance = $this->getValueInstance($key);
+            } else {
+                $instance = $this->resolveInstance($key);
+            }
+        } catch (\ReflectionException $e) {
+            throw new NotFoundException(sprintf('Key "%s" could not be resolved. ', $key));
```
Comment (lordrhodos) [inline src/Container/Container.php:128]:
```suggestion
            throw new NotFoundException(\sprintf('Key "%s" could not be resolved.', $key));
```

File: src/Container/Container.php [hunk 5]
```diff
@@ -115,8 +175,12 @@ public function getNew($key)
      *
      * @param string $key
      */
-    public function destroyInstance($key)
+    public function destroyInstance($key): void
```
Comment (lordrhodos) [inline src/Container/Container.php:178]:
```suggestion
    public function destroyInstance(string $key): void
```

and phpdoc can be removed

File: src/RouterInterface.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
+<?php
+/**
+ * Starlit App.
+ *
+ * @copyright Copyright (c) 2016 Starweb AB
```
Comment (lordrhodos) [inline src/RouterInterface.php:5]:
```suggestion
 * @copyright Copyright (c) 2019 Starweb AB
```

File: src/RouterInterface.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    public function clearRoutes(): void;
+
+    public function getRoutes(): RouteCollection;
+
+    /**
+     * Resolve controller action and return callable properties.
+     *
+     * @throws ResourceNotFoundException
+     */
+    public function route(Request $request): AbstractController;
```
Comment (lordrhodos) [inline src/RouterInterface.php:37]: This is a good start, but should be better implmented against a contract (interface)

File: src/ViewInterface.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+     */
+    public function __construct(array $options = []);
+
+    /**
+     * Set options.
+     */
+    public function setOptions(array $options): void;
+
+    public function setLayout(string $script);
```
Comment (lordrhodos) [inline src/ViewInterface.php:20]: do we have a mixed return value here or can we add a return type declaration?

File: src/ViewInterface.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    public function escape(?string $string, int $flags = ENT_QUOTES): string;
+
+    /**
+     * Returns view variable escaped for view script output.
+     */
+    public function getEscaped(string $name): string;
+
+    public function addHelperClass(string $helperName, string $className): void;
+
+    public function getHelper(string $helperName): AbstractViewHelper;
```
Comment (lordrhodos) [inline src/ViewInterface.php:39]: same here, use an interface instead of the abstract class

**Review by lordrhodos — CHANGES_REQUESTED** [review]
I think we need to add interfaces for the abstract classes we use in return type declarations. Apart from that I see only smaller changes that need to be made.

**Review by PaulPylkas — COMMENTED** [review]
I think it might be better to modify `Container::resolveParameters` to be public and take an optional `array $overrideParameters` parameter. We can then send in `$this->request->attributes->all()` in `AbstractController`.

It's pretty much just a more generalized implementation of the same thing. What do you think?

File: src/Container/Container.php [hunk 10]
```diff
@@ -36,76 +41,131 @@ class Container implements ContainerInterface
      * @param mixed  $value
      * @return Container
      */
-    public function set($key, $value)
+    public function set(string $key, $value): self
```
Comment (lordrhodos) [inline src/Container/Container.php:44]:
```suggestion
    public function set($key, $value): self
```

Comment (lordrhodos) [reply src/Container/Container.php:44]: @PaulPylkas can you check this, I think we need to remove the type declaration here or not?

**Comment by lordrhodos** [comment]
@PaulPylkas I have changed the visibility for `Container::resolveInstance` to `private` as we do not need to expose it anymore.

**Review by lordrhodos — APPROVED** [review]
@PaulPylkas @felixsand Approved from my side. Let's get this merged into `1.0.x-dev` and apply the 1.0.0 fixes to that branch in dedicated PRs

**Pull request merged**